Add ENABLE_PUSH flag in the Upgrade HTTP2-Settings header

hkwi opened the pull request:
Push flag required for the case the initial upgrade request triggered server push.

hkwi committed:
Add ENABLE_PUSH flag in the Upgrade HTTP2-Settings header

Push flag required for the case the initial upgrade request
triggered server push.

Lukasa reviewed (changes requested):
Thanks for this!

This patch is reasonable, but there are a bunch of tests failing that need to be updated to take account of this change. Do you mind doing that?

hkwi commented:
I'll look into.

hkwi committed:
make the http11 connection enable_push optional

If enable_push option was not explicitly set, HTTP2-Settings header
will not include ENABLE_PUSH option.

hkwi committed:
apply autopep8

fix for `flake8 --max-complexity 15`

Lukasa commented on hyper/http11/connection.py:
```diff
@@ -276,6 +277,9 @@ def _add_upgrade_headers(self, headers):
         http2_settings.settings[SettingsFrame.INITIAL_WINDOW_SIZE] = 65535
+        if self._enable_push is not None:
+            http2_settings.settings[
+                SettingsFrame.ENABLE_PUSH] = int(self._enable_push)
```
Rather than break here, can you do this?

```python
http2_settings.settings[SettingsFrame.ENABLE_PUSH] = (
    int(self._enable_push)
)
```

hkwi replied:
Yes.

Lukasa reviewed (changes requested):
So you've fixed some of the test failures, but you're now not actually testing that this data gets emitted. We do need to add a new test to cover this branch.

hkwi committed:
change code formatting

Lukasa reviewed (changes requested):
So, as noted, this is not really tested in any sense right now. It would be good to see the appropriate testing changes made.

hkwi committed:
add server push test with h2c upgrade

hkwi committed:
apply autopep8

hkwi committed:
apply autopep8

hkwi committed:
apply autopep8

hkwi committed:
fix typo

hkwi commented:
One travis-ci job is failing unexpectedly with empty raw log. Can I ask help?

Lukasa commented:
I've restarted that build. =)

hkwi committed:
fix test synchronization

Some tests requires synchronization to let the connection
state machine working.

hkwi committed:
remove unused variable

hkwi commented:
I added http11 h2c upgrading test by inheriting TestServerPush which is for http20 h2, and test passed.

Lukasa reviewed (changes requested):
So, I'm not sure about this approach.

You've added `getpushes` to the `HTTP11Connection` object with the goal of having it be a dummy method for catching the upgrade. That's understandable, but I'm not sure that this is the way users should be interacting with the abstract method: in particular because if the connection *does not* upgrade, which the user cannot guarantee will not happen, the response object is lost.

For the purpose of testing it's probably enough to call `getresponse` and then check for pushes. How does that sound?

hkwi commented:
I agree that `HTTP11Connection.getpushes` is a little bit hacky. This was added to meet the strict tests of `test_promise_before_headers` or `test_capture_all_promises`. If we don't need to satisfy these tests for http11 upgrade, implementation would be more relaxed style. I'll create another PR for that version later.
BTW, the response will be stored in `self._response` for later `self.get_response` call, and would not lost.

Lukasa commented:
I think those tests don't need to work for upgrade, but we should write a test that *does* work as expected (e.g. you get a HTTP/2 response and *then* look for pushes on that stream)
